README.md
# obs-picamera

[![PyPI](https://img.shields.io/pypi/v/obs-picamera?style=flat-square)](https://pypi.python.org/pypi/obs-picamera/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/obs-picamera?style=flat-square)](https://pypi.python.org/pypi/obs-picamera/)
[![PyPI - License](https://img.shields.io/pypi/l/obs-picamera?style=flat-square)](https://pypi.python.org/pypi/obs-picamera/)
[![Coookiecutter - Wolt](https://img.shields.io/badge/cookiecutter-Wolt-00c2e8?style=flat-square&logo=cookiecutter&logoColor=D4AA00&link=https://github.com/woltapp/wolt-python-package-cookiecutter)](https://github.com/woltapp/wolt-python-package-cookiecutter)


---

**Documentation**: [https://gluap.github.io/obs-picamera](https://gluap.github.io/obs-picamera)

**Source Code**: [https://github.com/gluap/obs-picamera](https://github.com/gluap/obs-picamera)

**PyPI**: [https://pypi.org/project/obs-picamera/](https://pypi.org/project/obs-picamera/)

---

A python module to record video of overtaking events to allow vehicle type identification for OpenBikeSensor overtaking events.

## Building the device

This software should work with a Pi Zero W or Pi Zero 2 with raspberry pi camera independently from the case it's mounted in.

The idea is to run this on a raspberry pi zero case which can be somehow mounted on the bicycle. An example case for such a mounting with a list of components used can be found [here](https://github.com/gluap/OpenBikeSensor3dPrintableCase/tree/camlid/src/Attachments), compiled STL files to match it are available [with logos](https://github.com/gluap/OpenBikeSensor3dPrintableCase/tree/camlid/export/logo/OpenBikeSensor) [and without](https://github.com/gluap/OpenBikeSensor3dPrintableCase/tree/camlid/export/Attachments)

## Installation of ``obs-picamera``

The software is meant to run on a raspberry pi zero. I suggest installing it with Raspberry Pi os. As prerequisites the following packages are required:

```commandline
sudo apt install python-picamera2 python3-pip
```

The package can then be installed as usual via pip.

```sh
pip install --user obs-picamera # add --upgrade if you want to upgrade
```

## Using ``obs-picamera``

You can run the program by calling (``pip install --user`` has installed the binary in your ``~/.local/bin/``)
```commandline
obs_picamera
```

When it finds an OpenBikeSensor via bluetooth enabled, it will pair with it and whenever it receives an overtaking event
- Events for one OpenBikeSensor Track are stored in a directory with the track-id as the directory name.
- For each event a short ``h264`` video file is saved, the filename matching the system time of the OpenBikeSensor.
- Next to the video file a ``.json`` file with the Data from the overtaking event is placed. Its content is self-explanatory. The distance already has the handlebar width deducted.

I suggest starting obs_picamera automatically at boot - for instance via crontab entry and switching the pi on while still in the home WIFI (which will enable it to pick up a sensible time). Or one may create a phone wifi for it to pick up the time and date



## Development

* Clone this repository
* Requirements:
  * [Poetry](https://python-poetry.org/)
  * Python 3.7+
* Create a virtual environment and install the dependencies

```sh
poetry install
```

* Activate the virtual environment

```sh
poetry shell
```

### Testing

```sh
pytest
```

### Documentation

The documentation is automatically generated from the content of the [docs directory](./docs) and from the docstrings
 of the public signatures of the source code. The documentation is updated and published as a [Github project page
 ](https://pages.github.com/) automatically as part each release.

### Releasing

Trigger the [Draft release workflow](https://github.com/gluap/obs-picamera/actions/workflows/draft_release.yml)
(press _Run workflow_). This will update the changelog & version and create a GitHub release which is in _Draft_ state.

Find the draft release from the
[GitHub releases](https://github.com/gluap/obs-picamera/releases) and publish it. When
 a release is published, it'll trigger [release](https://github.com/gluap/obs-picamera/blob/master/.github/workflows/release.yml) workflow which creates PyPI
 release and deploys updated documentation.

### Pre-commit

Pre-commit hooks run all the auto-formatters (e.g. `black`, `isort`), linters (e.g. `mypy`, `flake8`), and other quality
 checks to make sure the changeset is in good shape before a commit/push happens.

You can install the hooks with (runs for each commit):

```sh
pre-commit install
```

Or if you want them to run only for each push:

```sh
pre-commit install -t pre-push
```

Or if you want e.g. want to run all checks manually for all files:

```sh
pre-commit run --all-files
```

---

This project was generated using the [wolt-python-package-cookiecutter](https://github.com/woltapp/wolt-python-package-cookiecutter) template.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2023-02-04
### Added
- Rudimentary GUI
- Save distance history with video
- Run via Fastapi

## [0.1.0] - 2023-01-31
### Added
- Documentation

## [0.0.5] - 2023-01-31
### Fixed
- re-introduce py39 (as raspbian has no 310 yet.)
- better camera settings, ring buffer tested.

## [0.0.4] - 2023-01-30
### Fixed
- File handling from command line script
- Use of non-existant loop
- Rename command line scripts

## [0.0.3] - 2023-01-30
### Added
- Bluetooth
- Unittests
- Command line script

## [0.0.2] - 2023-01-28
### Added
- Set up github workflows and created pypi package

[Unreleased]: https://github.com/gluap/obs-picamera/compare/0.2.0...master
[0.2.0]: https://github.com/gluap/obs-picamera/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/gluap/obs-picamera/compare/0.0.5...0.1.0
[0.0.5]: https://github.com/gluap/obs-picamera/compare/0.0.4...0.0.5
[0.0.4]: https://github.com/gluap/obs-picamera/compare/0.0.3...0.0.4
[0.0.3]: https://github.com/gluap/obs-picamera/compare/0.0.2...0.0.3
[0.0.2]: https://github.com/gluap/obs-picamera/tree/0.0.2



tests/test_api.py
import sys
from unittest.mock import MagicMock

sys.modules["picamera2"] = MagicMock()
sys.modules["picamera2.encoders"] = MagicMock()
sys.modules["picamera2.outputs"] = MagicMock()

from fastapi.testclient import TestClient  # noqa E402 # isort: skip
from obs_picamera.api import app  # noqa E402 # isort: skip

client = TestClient(app)


def test_app() -> None:
    r = client.get("/v1/preview.jpeg")
    client.get("/v1/history")
    client.get("/v1/state")

    assert r.status_code == 200


tests/test_dummy.py
import obs_picamera
def test_dummy():
    assert True

tests/test_recorder.py
import pathlib
import sys
from unittest.mock import MagicMock

sys.modules["picamera2"] = MagicMock()
sys.modules["picamera2.encoders"] = MagicMock()
sys.modules["picamera2.outputs"] = MagicMock()

import obs_picamera.recorder as rec  # noqa E402 # isort: skip


def test_recorder_init(tmp_path: pathlib.Path) -> None:
    recorder = rec.Recorder()
    fp = open(tmp_path / "fck", "wb")
    recorder.save_snippet_to(fp)
    recorder.jpeg_screenshot()
    assert recorder.output.fileoutput == fp


src/obs_picamera/api.py
import asyncio
import json
import logging
import os
from pathlib import Path

import toml
from fastapi import FastAPI, Response

from obs_picamera.bluetooth import ObsBT, ObsScannerError
from obs_picamera.recorder import Recorder

logging.basicConfig(level=logging.WARNING)
logger = logging.getLogger()

recorder = Recorder()


# flake8: noqa
def main() -> FastAPI:
    app = FastAPI()

    obsbt = ObsBT()

    @app.get("/v1/preview.jpeg")
    async def preview() -> Response:
        return Response(recorder.jpeg_screenshot(), media_type="image/jpeg")

    def record_callback(**kwargs) -> None:  # type: ignore # pragma: no cover
        target_dir = Path("recordings") / kwargs["track_id"]
        target_dir.mkdir(parents=True, exist_ok=True)
        recorder.save_snippet_to(
            open(target_dir / f"{kwargs['sensortime']}.h264", "wb")
        )
        toml.dump(
            {str(target_dir / f"{kwargs['sensortime']}.h264"): kwargs},
            open(target_dir / "events.toml", "a"),
        )
        logger.info(f"saved to {target_dir}")

    obsbt.overtaking_callbacks.append(record_callback)

    async def bt_loop() -> None:  # pragma: no cover
        while True:
            try:
                await obsbt.run()
            except ObsScannerError:
                logger.exception("Restarting bluetooth connection")

    @app.on_event("startup")
    async def start_loop() -> None:
        logger.info("starting loop")
        asyncio.create_task(bt_loop())

    @app.get("/v1/history")
    async def history() -> Response:
        return Response(json.dumps(list(obsbt.history)), media_type="application/json")

    @app.get("/v1/state")
    async def state() -> Response:
        [time, left, right] = obsbt.history[-1] if len(obsbt.history) > 0 else [0, 0, 0]
        return Response(
            json.dumps(
                {
                    "sensortime": time,
                    "distance_l": left,
                    "distance_r": right,
                    "track": obsbt.track_id,
                    "handlebar_size": (obsbt.handlebar_left, obsbt.handlebar_right),
                    "connected": obsbt.bt_connected,
                }
            ),
            media_type="application/json",
        )

    @app.get("/v1/tracks")
    async def tracks() -> Response:
        dirs = os.listdir("recordings")
        return Response(json.dumps(dirs))

    @app.get("/v1/tracks/{id}")
    async def track(id: str) -> Response:
        events_file = Path(f"recordings/{id}") / "events.toml"
        res = toml.load(events_file.open("r"))
        return Response(json.dumps(res), media_type="application/json")

    return app


app = main()


src/obs_picamera/recorder.py
from io import BytesIO
from typing import BinaryIO

from picamera2 import Picamera2  # type: ignore
from picamera2.encoders import H264Encoder, Quality  # type: ignore
from picamera2.outputs import CircularOutput  # type: ignore


class Recorder:
    """
    Recorder for picamera that starts recording to a ring buffer when initialized
    and allows to save the last 5 seconds of video from the buffer.
    """

    def __init__(self) -> None:
        self.picam2 = Picamera2()
        fps = 25
        dur = 10
        micro = int((1 / fps) * 1000000)
        vconfig = self.picam2.create_video_configuration(
            main={"size": (800, 606)}, lores={"size": (800, 606)}, encode="lores"
        )
        vconfig["controls"]["FrameDurationLimits"] = (micro, micro)
        self.picam2.configure(vconfig)
        self.encoder = H264Encoder(
            iperiod=1,
            repeat=True,
        )
        self.output = CircularOutput(
            buffersize=int(fps * (dur + 0.2)), outputtofile=False
        )
        self.picam2.start_recording(self.encoder, self.output, Quality.HIGH)

    def save_snippet_to(self, fp: BinaryIO) -> None:
        """
        Save the last 5 seconds of video from the buffer to the filename
        :param fp:
        :return:
        """
        self.output.fileoutput = fp
        self.output.stop()

    def jpeg_screenshot(self) -> bytes:
        data = BytesIO()
        self.picam2.capture_file(data, format="jpeg")
        return data.getvalue()

    def __del__(self) -> None:
        self.picam2.stop_recording()


src/obs_picamera/bluetooth.py
import asyncio
import logging
import struct
import sys
import time
from collections import deque
from typing import Callable, List, NamedTuple, Optional

import bleak
from bleak.exc import BleakError

SENSOR_DISTANCE_NOTIFICATION_UUID = "1FE7FAF9-CE63-4236-0004-000000000002"
CLOSE_PASS_NOTIFICATION_UUID = "1FE7FAF9-CE63-4236-0004-000000000003"
HANDLEBAR_OFFSET_UUID = "1FE7FAF9-CE63-4236-0004-000000000004"
TRACK_ID_UUID = "1FE7FAF9-CE63-4236-0004-000000000005"

logger = logging.getLogger(__name__)
#
# Blutooth scanner config
#

timeout_seconds: float = 15


class DistanceMeasurement(NamedTuple):
    sensortime: int
    distance_l: int
    distance_r: int


class ObsScanner:
    unittesting: bool = False

    def __init__(self) -> None:
        if sys.version_info < (3, 10):  # pragma: no cover
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)

        self._scanner = bleak.BleakScanner(detection_callback=self.detection_callback)
        self.obs_address: str | None = None
        self.scanning = asyncio.Event()

    def detection_callback(
        self, device: bleak.BLEDevice, advertisement_data: bleak.AdvertisementData
    ) -> None:
        if device.name.startswith("OpenBikeSensor"):
            logger.info(f"Found OpenBikeSensor {device}")
            self.obs_address = device.address
            self.scanning.clear()

    async def run(self) -> None:
        logger.info("Scan for devices")
        if not self.unittesting:
            self.obs_address = None
        await self._scanner.start()
        self.scanning.set()
        end_time = time.time() + timeout_seconds
        while self.scanning.is_set():
            if time.time() > end_time:
                self.scanning.clear()
                logger.info("Scan has timed out so we terminate")
            await asyncio.sleep(0.1)
        await self._scanner.stop()
        if self.obs_address is None:
            raise ObsScannerError("No OBS found")  # flake8: noqa


class ObsBT:
    def __init__(self, scanner: Optional[ObsScanner] = None) -> None:
        self.bt_connected: bool = False
        self.obs_address: str | None = None
        self.my_scanner: ObsScanner = scanner if scanner is not None else ObsScanner()
        self.handlebar_left: int = 0
        self.handlebar_right: int = 0
        self.track_id: str | None = None
        self.overtaking_callbacks: List[Callable] = []
        self.recording_callbacks: List[Callable] = []
        self.bt_connecting: bool = False
        self.unittesting: bool = False
        self.history: deque = deque(maxlen=100)

    async def find_obs(self) -> None:
        await self.my_scanner.run()
        if self.obs_address is None:
            self.obs_address = self.my_scanner.obs_address
        assert self.obs_address is not None

    def overtaking_handler(self, sender: str, data: bytearray) -> None:
        """Simple notification handler which prints the data received."""
        t, l, r = struct.unpack("Ihh", data)
        logger.info(
            f"Overtaking event: sensortime: {t}, Left distance {l}, right distance {r}"
        )
        for callback in self.overtaking_callbacks:
            callback(
                distance_l=l - self.handlebar_left,
                distance_r=r - self.handlebar_right,
                handlebars_l=self.handlebar_left,
                handlebar_r=self.handlebar_right,
                track_id=self.track_id,
                sensortime=t,
                history=list(self.history),
            )

    def history_handler(self, sender: str, data: bytearray) -> None:
        t, l, r = struct.unpack("Ihh", data)
        self.history.append(
            DistanceMeasurement(t, l - self.handlebar_left, r - self.handlebar_right)
        )

        for callback in self.recording_callbacks:
            callback(
                distance_l=l - self.handlebar_left,
                distance_r=r - self.handlebar_right,
                handlebars_l=self.handlebar_left,
                handlebar_r=self.handlebar_right,
                track_id=self.track_id,
                sensortime=t,
            )

    async def connect(self) -> None:
        def disconnected_callback(client: bleak.BleakClient) -> None:
            logger.info("DISCONNECTED")
            self.bt_connected = False

        async with bleak.BleakClient(
            self.obs_address, disconnected_callback=disconnected_callback, timeout=25  # type: ignore
        ) as client:
            logger.info(f"Connected: {client.is_connected}")
            await client.start_notify(
                CLOSE_PASS_NOTIFICATION_UUID, self.overtaking_handler
            )
            await client.start_notify(
                SENSOR_DISTANCE_NOTIFICATION_UUID, self.history_handler
            )
            self.track_id = (await client.read_gatt_char(TRACK_ID_UUID)).decode("utf-8")
            handlebar = await client.read_gatt_char(HANDLEBAR_OFFSET_UUID)
            self.handlebar_left, self.handlebar_right = struct.unpack("hh", handlebar)
            logger.info(
                f"track id is {self.track_id}, handlebar: <{self.handlebar_left} | {self.handlebar_right} >"
            )
            self.bt_connected = True
            self.bt_connecting = False
            while True:
                if not self.bt_connected or self.unittesting:
                    break
                await asyncio.sleep(1)

    async def run(self) -> None:
        await self.find_obs()
        while not self.bt_connected:
            try:
                if not self.bt_connecting:
                    logger.info(f"connecting to {self.obs_address}")
                    self.bt_connecting = True
                    assert isinstance(self.obs_address, str)
                    await self.connect()
                else:  # pragma: no cover
                    await asyncio.sleep(1)
            except BleakError:  # pragma: no cover
                logger.info(
                    f"lost connection, reconnecting. connecting={self.bt_connecting}"
                )


class ObsScannerError(Exception):
    pass


if __name__ == "__main__":  # pragma: no cover
    while True:
        try:
            obsbt = ObsBT()
            asyncio.run(obsbt.run())
        except (ObsScannerError, bleak.BleakError):
            logger.exception("ERROR. Retry")


src/obs_picamera/scripts/ui.py
# type: ignore
# flake8: noqa
"""
   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
"""

import io
import json
import logging
import os
import threading
import time
import traceback

import remi
import requests

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger()

import remi.gui as gui
from remi import App, start

API_HOST = (
    "http://localhost:5000" if "API_HOST" not in os.environ else os.environ["API_HOST"]
)


class RefreshImageWidget(remi.gui.Image):
    def __init__(self, url=None, **kwargs):
        self.app_instance = None
        super().__init__("/res:logo.jpg", **kwargs)
        self.frame_index = 0
        self._buf = None
        self.url = url

    def search_app_instance(self, node):
        if issubclass(node.__class__, remi.server.App):
            return node
        if not hasattr(node, "get_parent"):
            return None
        return self.search_app_instance(node.get_parent())

    def load(self):
        try:
            data = requests.get(self.url)
            self._buf = io.BytesIO(data.content)
            self.refresh()
        except requests.ConnectionError:
            logger.info("couldn't reload")

    def refresh(self, *args):
        if self.app_instance == None:
            self.app_instance = self.search_app_instance(self)
            if self.app_instance == None:
                return
        self.frame_index = self.frame_index + 1
        self.app_instance.execute_javascript(
            """
            url = '/%(id)s/get_image_data?index=%(frame_index)s';

            xhr = null;
            xhr = new XMLHttpRequest();
            xhr.open('GET', url, true);
            xhr.responseType = 'blob'
            xhr.onload = function(e){
                urlCreator = window.URL || window.webkitURL;
                urlCreator.revokeObjectURL(document.getElementById('%(id)s').src);
                imageUrl = urlCreator.createObjectURL(this.response);
                document.getElementById('%(id)s').src = imageUrl;
            }
            xhr.send();
            """
            % {"id": id(self), "frame_index": self.frame_index}
        )

    def get_image_data(self, index=0):
        try:
            self._buf.seek(0)
            headers = {"Content-type": "image/jpeg"}
            return [self._buf.read(), headers]
        except:
            print(traceback.format_exc())
        return None, None


class MyApp(App):
    def __init__(self, *args, recorder=None):
        super().__init__(*args)

    def idle(self):
        if hasattr(self, "data_table"):
            i = 0
            for k, v in self.data.items():
                i += 1
                self.data_table.item_at(i, 0).set_text(str(k))
                self.data_table.item_at(i, 1).set_text(str(v))
        else:
            logger.info("no counter")

    def _get_list_from_app_args(self, name):
        return super()._get_list_from_app_args(name)

    def preview_screen(self) -> gui.Widget:
        self.to_review = gui.Button("review tracks")
        tabs = gui.HBox(
            children=[self.to_review],
            style={"margin": "0px auto", "background-color": "lightgray"},
        )

        self.stop_flag = False

        self.img = RefreshImageWidget(
            f"{API_HOST}/v1/preview.jpeg", margin="10px", width=0, height=0
        )
        self.video = gui.VideoPlayer(width=806, height=600)
        self.check = gui.CheckBoxLabel(
            "Preview", False, width=200, height=30, margin="10px"
        )

        self.data_table = gui.TableWidget(8, 2)
        self.data_table.item_at(0, 0).set_text("k")
        self.data_table.item_at(0, 1).set_text("v")

        self.check.onchange.do(self.on_check_change)
        video_container = gui.VBox(
            [tabs, self.img, self.check, self.data_table], margin="8pt auto"
        )

        self.refresh = False
        self.running = True
        return video_container

    def review_screen(self) -> gui.Widget:
        self.to_preview = gui.Button("change page")
        reviewcontainer = gui.HBox(
            children=[self.to_preview],
            style={"margin": "0px auto", "background-color": "lightgray"},
        )
        return reviewcontainer

    def main(self):
        self.data = {}
        # self.recorder = Recorder()

        self.previewcontainer = self.preview_screen()
        self.reviewcontainer = self.review_screen()
        self.to_review.onclick.do(self.set_different_root_widget, self.reviewcontainer)
        self.to_preview.onclick.do(
            self.set_different_root_widget, self.previewcontainer
        )

        t = threading.Thread(target=self.update)
        t.start()

        return self.reviewcontainer

    def set_different_root_widget(self, emitter, page_to_be_shown):
        self.set_root_widget(page_to_be_shown)
        self.check.set_value(False)
        self.refresh = False

    def update_data(self):
        try:
            current = requests.get(f"{API_HOST}/v1/state")
            self.data = json.loads(current.content)
        except requests.RequestException:
            pass

    def update(self):
        lasttime: int = 0
        while self.running:
            time.sleep(0.1)
            if self.refresh and lasttime != int(time.time()):
                self.img.load()
                lasttime = int(time.time())
            self.update_data()

    def __main(self):
        # the margin 0px auto centers the main container
        verticalContainer = gui.Container(
            width=540,
            margin="0px auto",
            style={"display": "block", "overflow": "hidden"},
        )

        horizontalContainer = gui.Container(
            width="100%",
            layout_orientation=gui.Container.LAYOUT_HORIZONTAL,
            margin="0px",
            style={"display": "block", "overflow": "auto"},
        )

        subContainerLeft = gui.Container(
            width=320,
            style={"display": "block", "overflow": "auto", "text-align": "center"},
        )
        self.img = gui.Image("/res:logo.png", height=100, margin="10px")
        self.img.onclick.do(self.on_img_clicked)

        self.table = gui.Table.new_from_list(
            [
                ("ID", "First Name", "Last Name"),
                ("101", "Danny", "Young"),
                ("102", "Christine", "Holand"),
                ("103", "Lars", "Gordon"),
                ("104", "Roberto", "Robitaille"),
                ("105", "Maria", "Papadopoulos"),
            ],
            width=300,
            height=200,
            margin="10px",
        )
        self.table.on_table_row_click.do(self.on_table_row_click)

        # the arguments are	width - height - layoutOrientationOrizontal
        subContainerRight = gui.Container(
            style={
                "width": "220px",
                "display": "block",
                "overflow": "auto",
                "text-align": "center",
            }
        )
        self.count = 0
        self.counter = gui.Label("", width=200, height=30, margin="10px")

        self.lbl = gui.Label("This is a LABEL!", width=200, height=30, margin="10px")

        self.bt = gui.Button("Press me!", width=200, height=30, margin="10px")
        # setting the listener for the onclick event of the button
        self.bt.onclick.do(self.on_button_pressed)

        self.txt = gui.TextInput(width=200, height=30, margin="10px")
        self.txt.set_text("This is a TEXTAREA")
        self.txt.onchange.do(self.on_text_area_change)

        self.spin = gui.SpinBox(1, 0, 100, width=200, height=30, margin="10px")
        self.spin.onchange.do(self.on_spin_change)

        self.progress = gui.Progress(1, 100, width=200, height=5)

        self.check = gui.CheckBoxLabel(
            "Label checkbox", True, width=200, height=30, margin="10px"
        )
        self.check.onchange.do(self.on_check_change)

        self.btInputDiag = gui.Button(
            "Open InputDialog", width=200, height=30, margin="10px"
        )
        self.btInputDiag.onclick.do(self.open_input_dialog)

        self.btFileDiag = gui.Button(
            "File Selection Dialog", width=200, height=30, margin="10px"
        )
        self.btFileDiag.onclick.do(self.open_fileselection_dialog)

        self.btUploadFile = gui.FileUploader("./", width=200, height=30, margin="10px")
        self.btUploadFile.onsuccess.do(self.fileupload_on_success)
        self.btUploadFile.onfailed.do(self.fileupload_on_failed)

        items = ("Danny Young", "Christine Holand", "Lars Gordon", "Roberto Robitaille")
        self.listView = gui.ListView.new_from_list(
            items, width=300, height=120, margin="10px"
        )
        self.listView.onselection.do(self.list_view_on_selected)

        self.link = gui.Link(
            "http://localhost:8081",
            "A link to here",
            width=200,
            height=30,
            margin="10px",
        )

        self.dropDown = gui.DropDown.new_from_list(
            ("DropDownItem 0", "DropDownItem 1"), width=200, height=20, margin="10px"
        )
        self.dropDown.onchange.do(self.drop_down_changed)
        self.dropDown.select_by_value("DropDownItem 0")

        self.slider = gui.Slider(10, 0, 100, 5, width=200, height=20, margin="10px")
        self.slider.onchange.do(self.slider_changed)

        self.colorPicker = gui.ColorPicker(
            "#ffbb00", width=200, height=20, margin="10px"
        )
        self.colorPicker.onchange.do(self.color_picker_changed)

        self.date = gui.Date("2015-04-13", width=200, height=20, margin="10px")
        self.date.onchange.do(self.date_changed)

        self.video = gui.Widget(_type="iframe", width=290, height=200, margin="10px")
        self.video.attributes[
            "src"
        ] = "https://drive.google.com/file/d/0B0J9Lq_MRyn4UFRsblR3UTBZRHc/preview"
        self.video.attributes["width"] = "100%"
        self.video.attributes["height"] = "100%"
        self.video.attributes["controls"] = "true"
        self.video.style["border"] = "none"

        self.tree = gui.TreeView(width="100%", height=300)
        ti1 = gui.TreeItem("Item1")
        ti2 = gui.TreeItem("Item2")
        ti3 = gui.TreeItem("Item3")
        subti1 = gui.TreeItem("Sub Item1")
        subti2 = gui.TreeItem("Sub Item2")
        subti3 = gui.TreeItem("Sub Item3")
        subti4 = gui.TreeItem("Sub Item4")
        subsubti1 = gui.TreeItem("Sub Sub Item1")
        subsubti2 = gui.TreeItem("Sub Sub Item2")
        subsubti3 = gui.TreeItem("Sub Sub Item3")
        self.tree.append([ti1, ti2, ti3])
        ti2.append([subti1, subti2, subti3, subti4])
        subti4.append([subsubti1, subsubti2, subsubti3])

        # appending a widget to another, the first argument is a string key
        subContainerRight.append(
            [
                self.counter,
                self.lbl,
                self.bt,
                self.txt,
                self.spin,
                self.progress,
                self.check,
                self.btInputDiag,
                self.btFileDiag,
            ]
        )
        # use a defined key as we replace this widget later
        fdownloader = gui.FileDownloader(
            "download test", "../remi/res/logo.png", width=200, height=30, margin="10px"
        )
        subContainerRight.append(fdownloader, key="file_downloader")
        subContainerRight.append(
            [
                self.btUploadFile,
                self.dropDown,
                self.slider,
                self.colorPicker,
                self.date,
                self.tree,
            ]
        )
        self.subContainerRight = subContainerRight

        subContainerLeft.append(
            [self.img, self.table, self.listView, self.link, self.video]
        )

        horizontalContainer.append([subContainerLeft, subContainerRight])

        menu = gui.Menu(width="100%", height="30px")
        m1 = gui.MenuItem("File", width=100, height=30)
        m2 = gui.MenuItem("View", width=100, height=30)
        m2.onclick.do(self.menu_view_clicked)
        m11 = gui.MenuItem("Save", width=100, height=30)
        m12 = gui.MenuItem("Open", width=100, height=30)
        m12.onclick.do(self.menu_open_clicked)
        m111 = gui.MenuItem("Save", width=100, height=30)
        m111.onclick.do(self.menu_save_clicked)
        m112 = gui.MenuItem("Save as", width=100, height=30)
        m112.onclick.do(self.menu_saveas_clicked)
        m3 = gui.MenuItem("Dialog", width=100, height=30)
        m3.onclick.do(self.menu_dialog_clicked)

        menu.append([m1, m2, m3])
        m1.append([m11, m12])
        m11.append([m111, m112])

        menubar = gui.MenuBar(width="100%", height="30px")
        menubar.append(menu)

        verticalContainer.append([menubar, horizontalContainer])

        # this flag will be used to stop the display_counter Timer
        self.stop_flag = False

        # kick of regular display of counter
        self.display_counter()

        # returning the root widget
        return verticalContainer

    # listener function
    def on_img_clicked(self, widget):
        self.lbl.set_text("Image clicked!")

    def on_table_row_click(self, table, row, item):
        self.lbl.set_text("Table Item clicked: " + item.get_text())

    def on_button_pressed(self, widget):
        self.lbl.set_text("Button pressed! ")
        self.bt.set_text("Hi!")

    def on_text_area_change(self, widget, newValue):
        self.lbl.set_text("Text Area value changed!")

    def on_spin_change(self, widget, newValue):
        self.lbl.set_text("SpinBox changed, new value: " + str(newValue))

    def on_check_change(self, widget, newValue):
        self.refresh = newValue
        if self.refresh:
            self.img.set_size(806, 600)
        else:
            self.img.set_size(0, 0)

    def open_input_dialog(self, widget):
        self.inputDialog = gui.InputDialog(
            "Input Dialog", "Your name?", initial_value="type here", width=500
        )
        self.inputDialog.confirm_value.do(self.on_input_dialog_confirm)

        # here is returned the Input Dialog widget, and it will be shown
        self.inputDialog.show(self)

    def on_input_dialog_confirm(self, widget, value):
        self.lbl.set_text("Hello " + value)

    def open_fileselection_dialog(self, widget):
        self.fileselectionDialog = gui.FileSelectionDialog(
            "File Selection Dialog", "Select files and folders", False, "."
        )
        self.fileselectionDialog.confirm_value.do(self.on_fileselection_dialog_confirm)

        # here is returned the Input Dialog widget, and it will be shown
        self.fileselectionDialog.show(self)

    def on_fileselection_dialog_confirm(self, widget, filelist):
        # a list() of filenames and folders is returned
        self.lbl.set_text("Selected files: %s" % ",".join(filelist))
        if len(filelist):
            f = filelist[0]
            # replace the last download link
            fdownloader = gui.FileDownloader(
                "download selected", f, width=200, height=30
            )
            self.subContainerRight.append(fdownloader, key="file_downloader")

    def list_view_on_selected(self, widget, selected_item_key):
        """The selection event of the listView, returns a key of the clicked event.
        You can retrieve the item rapidly
        """
        self.lbl.set_text(
            "List selection: " + self.listView.children[selected_item_key].get_text()
        )

    def drop_down_changed(self, widget, value):
        self.lbl.set_text("New Combo value: " + value)

    def slider_changed(self, widget, value):
        self.lbl.set_text("New slider value: " + str(value))

    def color_picker_changed(self, widget, value):
        self.lbl.set_text("New color value: " + value)

    def date_changed(self, widget, value):
        self.lbl.set_text("New date value: " + value)

    def menu_save_clicked(self, widget):
        self.lbl.set_text("Menu clicked: Save")

    def menu_saveas_clicked(self, widget):
        self.lbl.set_text("Menu clicked: Save As")

    def menu_open_clicked(self, widget):
        self.lbl.set_text("Menu clicked: Open")

    def menu_view_clicked(self, widget):
        self.lbl.set_text("Menu clicked: View")

    def fileupload_on_success(self, widget, filename):
        self.lbl.set_text("File upload success: " + filename)

    def fileupload_on_failed(self, widget, filename):
        self.lbl.set_text("File upload failed: " + filename)

    def on_close(self):
        """Overloading App.on_close event to stop the Timer."""
        self.stop_flag = True
        self.running = False

        super().on_close()


def main():
    start(
        MyApp,
        debug=True,
        address="0.0.0.0",
        port=8081,
        start_browser=True,
        multiple_instance=True,
    )


if __name__ == "__main__":
    # starts the webserver
    # optional parameters
    # start(MyApp,address='127.0.0.1', port=8081, multiple_instance=False,enable_file_cache=True, update_interval=0.1, start_browser=True)
    main()
